### DISCUSSION TITLE
Questions about go-netbox/v3 (openapi)

### DISCUSSION THREAD

**Discussion by smutel** [netbox-community/go-netbox#166] [Q&A] [answered]
Hello,

Thanks for the job done to change this library to use openapi.

I have several questions:
* Could we have an example how to add an object in Netbox ? We have some example to list objects but not to create one.
* How to pass insecure option (to not check TLS cert) to NewConfiguration() ?

Thanks.

**Comment by jqueuniet** [comment, marked as answer]
I posted a creation example earlier in the OpenAPI discussion, but here it is for convenience:

```go
// Create an IP address
req := netbox.WritableIPAddressRequest{
	Address: address,
}
req.SetTenant(tenant.Id)
req.SetStatus("active")
if vrf != nil {
	req.SetVrf(vrf.Id)
}

ipAddress, resp, err := client.
	IpamApi.
	IpamIpAddressesCreate(ctx).
	WritableIPAddressRequest(req).
	Execute()
```

Here is another for a VM interface with slightly different calls:

```go
req := netbox.NewWritableVMInterfaceRequestWithDefaults()
req.SetName(fmt.Sprintf("net%d", vif.Index))
req.SetVirtualMachine(vm.Id)

if vif.MTU != 0 {
	req.SetMtu(vif.MTU)
}

nbvif, resp, err := client.
	VirtualizationAPI.
	VirtualizationInterfacesCreate(ctx).
	WritableVMInterfaceRequest(*req).
	Execute()
```

Skipping TLS certificate validation is done by tweaking the `http.Client` in the usual manner:

```go
configuration := netbox.NewConfiguration()
configuration.HTTPClient = &http.Client{
	Transport: &http.Transport{
		TLSClientConfig: &tls.Config{
			InsecureSkipVerify: true,
		},
	},
}

apiClient = netbox.NewAPIClient(configuration)
```

**Reply by smutel** [reply]
Thanks for all these examples.

I would like to get the answer of a device role creation through the API.
With tcpdump we can see the answer from Netbox:
```
17:11:45.581401 IP 127.0.0.1.8000 > 127.0.0.1.37568: Flags [.], ack 363, win 509, options [nop,nop,TS val 1123285057 ecr 1123285057], length 0
E..4Mz@.@..G.........@....@..........(.....
B..AB..A
17:11:45.586456 IP 127.0.0.1.8000 > 127.0.0.1.37568: Flags [P.], seq 1:470, ack 363, win 512, options [nop,nop,TS val 1123285062 ecr 1123285057], length 469
E..     M{@.@..q.........@....@................
B..FB..AHTTP/1.1 400 Bad Request
Content-Type: application/json
Vary: HX-Request, Cookie, origin
Allow: GET, POST, PUT, PATCH, DELETE, HEAD, OPTIONS
X-Request-ID: d8ec80ef-091a-4e52-b232-8480bf4fcbc3
API-Version: 3.7
X-Content-Type-Options: nosniff
Referrer-Policy: same-origin
Cross-Origin-Opener-Policy: same-origin
X-Frame-Options: SAMEORIGIN
Content-Length: 41
Server: Unit/1.31.1
Date: Tue, 05 Mar 2024 16:11:45 GMT

{"name":["This field may not be blank."]}
```

I would like to retrieve `{"name":["This field may not be blank."]}` to display it as error.
Here is the piece of code:
```
resourceCreated, http, err := client.DcimAPI.DcimDeviceRolesCreate(ctx).WritableDeviceRoleRequest(*newResource).Execute()
```

However the value of http.Request.Response is `nil` so I don't know how to get this message in my code ...

**Reply by jqueuniet** [reply]
The second return value from executors is already a response object, you do not need to search for it into the Request field.

I handle Netbox errors using a generic function, which looks like this:

```go
	ipAddress, resp, err := client.
		IpamAPI.
		IpamIpAddressesCreate(ctx).
		WritableIPAddressRequest(ipInput).
		Execute()
	if err != nil {
		return nil, neterr.RaiseNetboxError(resp, err)
	}
```

```go
func RaiseNetboxError(response *http.Response, err error) error {
	if err == nil {
		err = ErrUnknownNetboxError
	}

	if response == nil {
		return fmt.Errorf("netbox error with nil response: %w", err)
	}

	if response.StatusCode == 409 {
		return liberr.NewResourceExhaustedError(response.Request.URL.Path)
	}

	bodyBytes, readErr := io.ReadAll(response.Body)
	if readErr != nil {
		return fmt.Errorf("error reading Netbox body response: %w", readErr)
	}

	bodyString := string(bodyBytes)
	return fmt.Errorf(
		"error during a Netbox request: %w (url: %s, method: %s, body: %s)",
		err,
		response.Request.URL,
		response.Request.Method,
		bodyString,
	)
}
```

As you can see, the body should be available in `http.Body`, using your naming convention. But IMHO you should probably change that name, it can be confused quite easily with the `net/http` standard Go library.

**Reply by smutel** [reply]
Thanks a lot !!!

Have another issue ... I have this bug : https://github.com/netbox-community/go-netbox/issues/165 so I bypass the error if statuscode == 201.

Here is my piece of code to create a device role:
```
resourceCreated, response, err := client.DcimAPI.DcimDeviceRolesCreate(ctx).WritableDeviceRoleRequest(*newResource).Execute()
```

The device role is created correctly (when bypassing issue) but resourceCreated.GetId() return 0.
What I am doing wrong ? Is-it link to issue https://github.com/netbox-community/go-netbox/issues/165 ?

**Reply by jqueuniet** [reply]
Yes, issue #165 is a deserialization bug, at best the resource recovered will be an unititialized struct, at worst a nil pointer, and in both cases is unlikely to be of any use. Until the library is fixed, the only workaround I can see is to write your own deserialization using the `response.Body` content.

**Reply by smutel** [reply]
Are all netbox objects affected by this bug or only some of them ?

**Reply by jqueuniet** [reply]
Only some of them, I haven't had any issue on IPAM endpoints this far, but did found a similar issue in DCIM and virtualization.

**Reply by jacobsalmela** [reply]
> Yes, issue #165 is a deserialization bug, at best the resource recovered will be an unititialized struct, at worst a nil pointer, and in both cases is unlikely to be of any use. Until the library is fixed, the only workaround I can see is to write your own deserialization using the `response.Body` content.

Does it need to be fixed in netbox itself or the openapi code to generate the library?

Ran into this on creating device types with this library.

**Reply by jqueuniet** [reply]
I took some time to dig into it this morning and found what seems to be a similar bug on the Netbox project: https://github.com/netbox-community/netbox/issues/14953

So the fact those fields are required seems intended and their absence seems to be a bug.

This could be worked around by editing the OpenAPI spec before generating the library code in the meantime, at least until Netbox solves the issue on their side.

**Reply by smutel** [reply]
What do you to add something like this in these project:
* https://github.com/smutel/go-netbox/tree/main/patchs
* https://github.com/smutel/go-netbox/blob/59bcec29858a60e9e27a698ed13182425a2ae269/utils/netbox_generate_client#L75

The generated patch can be used until the issue is completely fixed in Netbox project which could be more or less long to have. This avoid to keep an issue in this project because the issue is still in the Netbox project.

**Reply by jqueuniet** [reply]
There's already a process to patch the spec, using the `scripts/fix-spec.py`. Given it basically reserializes the whole spec file, I'm not sure using a text-based workflow with patch is going to work very well with it, or it would have to be applied first, probably using a separate bash script. I guess it does use the same PyYAML parser as upstream though, so maybe it doesn't change things too much.

YAML tree manipulations do have the advantage of allowing complex workflows and being a lot more resistant to minor changes from upstream. As this Netbox bug probably affects other endpoints than the one talked about here given the upstream issue comments, maybe iterating over all responses and turning optional every root field with `*_count` would be a more comprehensive solution.

Feel free to open a PR either way, though I'm not a maintainer here, just answering what question I can since I've spent a fair amount of time dealing with OpenAPI generator-based Go clients at this point.

**Reply by smutel** [reply]
> Only some of them, I haven't had any issue on IPAM endpoints this far, but did found a similar issue in DCIM and virtualization.

For now, quite all DCIM objects have the same issues ... :sob:
https://github.com/smutel/terraform-provider-netbox/blob/UpdateForNetbox3.7/tmp/README.md
